azure-pipelines.yml

```diff
@@ -5,20 +5,51 @@ variables:
   DOTNET_SKIP_FIRST_TIME_EXPERIENCE: 1
   DOTNET_SDK_VERSION: '3.0.100-preview5-011568'
 
+resources:
+  containers:
+  - container: postgres_11
+    image: postgres:11
+    ports:
+    - 5432:5432
+  - container: postgres_10
+    image: postgres:10
+    ports:
+    - 5432:5432
+  - container: postgres_9
+    image: postgres:9
+    ports:
+    - 5432:5432
+
 jobs:
 
 - job: Linux
 
   pool:
     vmImage: 'ubuntu-16.04'
 
+  strategy:
+    matrix:
+      postgres_11:
+        postgres_service: postgres_11
+      postgres_10:
+        postgres_service: postgres_10
+      postgres_9:
+        postgres_service: postgres_9
+
+  services:
+    postgres: $[variables['postgres_service']]
+
   steps:
   - bash: |
       export TRUNCATED_SHA1=${BUILD_SOURCEVERSION:0:9}
       echo "##vso[task.setvariable variable=truncated_sha1]$TRUNCATED_SHA1";
     displayName: 'Assign truncated SHA1 to environment variable'
 
-  - bash: sudo sh .build/setup_postgres.sh
+  - bash: |
+      docker exec -d $[variables['postgres_service']] apt-get -y install postgresql-11-postgis-2.5
```

Comment:
Just a thought @austindrenski: I wonder if getting PG via docker can't speed up our build regardless of multi-version testing. I can see that PostgreSQL installation takes around 1 minute 40 seconds in current builds, maybe this could be sped up?

Then again, we'd have wire up all the proper extensions/configuration (PostGIS, SSL, prepared transactions...), not sure it's worth the bother.

Reply:
I've had the same thought. In the long run, I think it's worthwhile to at least kick the tires here.

Reply:
The docker startup time ~35 seconds. So, definitely some value here.

For now, we're only getting this benefit on the Linux job, as it appears that AP doesn't currently support using Linux service container from a Windows agent job. But we might get to sidestep that altogether by building the jobs within containers too. Further testing needed, though.

Take a look at [build 149](https://dev.azure.com/npgsql/EFCore.PG/_build/results?buildId=149) for a good example. It failed, but only because our test suite isn't super version-aware at the moment.

Reply:
Going the Docker route, the Linux build time [drops](https://dev.azure.com/npgsql/EFCore.PG/_build/results?buildId=166) to 5:46 of which 4:20 was test execution. Whereas the windows side finishes in 12:08 of which 8:52 was test execution.

It doesn't seem like there's an off-the-shelf image for what we need (e.g. Windows + PostgreSQL + PostGIS), so containerizing the whole thing might require us to layer our own image based on something like https://github.com/appropriate/docker-postgis. 

I think there's some upside to exploring this further, including future proofing us against future CI changes, and making it easier to reproduce the exact CI environment locally. 

_(Still all conditional on the maintenance/usefulness trade-off, and our ability to avoid building all of the images from scratch.)_